### ISSUE TITLE
New behavior of `nmstatectl service` breaks backwards compatibility

### ISSUE THREAD

**Issue by drawks** [nmstate/nmstate#2539]
## Describe the bug
As of the merge of  https://github.com/nmstate/nmstate/commit/dbb98e1d30c727b5e47ea7daacc8f2936ab5a0e0 the behavior of `nmstatectl service` is to store a checksum of the applied configs and not reapply configs that have previously been applied.

This is a breaking change as you can no longer evaluate the contents of the `.applied` file to determine what the current running state should be. The semantics of the behavior of the command have fundamentally changed as well as the contents/format of the files involved in the service. This should not have happened in a patch release (behavior is changed between 2.2.21 and 2.2.22), typically a change of this nature would be implemented in a manner that preserves backwards compatibility and would be in a minor release and removal of backwards compatibility would be noted as a coming deprecation/breaking change to be subsequently included in a major release.

## Nmstate version
- Version 2.2.23

## NetworkManager version
- Version 1.44.0

## Platform
- Platform: rhel 9.3

## How reproducible
- 100%

## Steps to Reproduce
1. create a file `/etc/nmstate/somechanges.yml` with some changes to be made
2. run `sudo nmstatectl service`
3. observe that:
    1. `/etc/nmstate/somechanges.yml` is still present unlike in previous versions
    2. `/etc/nmstate/somechanges.applied` file contains a sha256 digest instead of being a full copy of the applied change as in previous versions
4. any software integration that depended on the previous documented behavior is now broken in unanticipated ways

### Additional context
The motivation for the change is documented as `At some envs like openshift machine-config-operator moving configuration
files can break in place upgrades` which acknowledges that the existing behavior exists and the change fundamentally changes how this command behaves. It is inappropriate and unexpected that a downstream consumer of the nmstate failing to work correctly within the documented behavior of the software would be enough to motivate a breaking change in a patch release that is subsequently pushed out to a stable release of an enterprise OS distribution. @qinqon as an employee of Redhat  I'm surprised to see that you signed off on this change

Also, it should be noted that the accepted change does not include any update to the manpage for `nmstatectl` which still says:
```
       service
              Apply all network state files ending with .yml in specified( default: /etc/nmstate) folder.   The  applied  network
              state file will be renamed with postfix .applied to prevent repeated applied on next run.
```
and [the changelog ](https://github.com/nmstate/nmstate/blob/base/CHANGELOG) fails to not that any breaking changes were included in the release instead declaring the changed behavior as a "bugfix" even though the original behavior was documented and behaved exactly as documented. Typically a "bug" is a behavior that varies from the expected and documented behavior.

**Comment by cathay4t** [comment]
I guess I can introduce a config in `/etc/nmstate/nmstate.conf` to use this new hash method:

```
[service]
remove_applied_yaml = false
```

Then by default, we still remove the applied YAML file and openshift machine-config-operator could use above config file.

@qinqon Is that OK to you?

**Comment by cathay4t** [comment]
@drawks Are you seeking nmstate next release to change back to old behavior or expecting us to change the `CHANGLOG.md` and manpage?

**Comment by qinqon** [comment]
@drawks we missed to document the change, although the .applied files were an implementation detail to prevent re-apply more than an nmtate interface.

There are other mechanism to check if the state was applied, like checking the `nmstatectl show` or if the nmstate.service unit was successful, but we are ok going back to old behaviour controller by a flag.

**Comment by cathay4t** [comment]
The manpage will be fixed by https://github.com/nmstate/nmstate/pull/2541

**Comment by drawks** [comment]
> I guess I can introduce a config in `/etc/nmstate/nmstate.conf` to use this new hash method:
> 
> ```
> [service]
> remove_applied_yaml = false
> ```
> 
> Then by default, we still remove the applied YAML file and openshift machine-config-operator could use above config file.
> 
> @qinqon Is that OK to you?

Yes, the new behavior should be configurable and it should not be the default behavior.

**Comment by drawks** [comment]
> @drawks we missed to document the change, although the .applied files were an implementation detail to prevent re-apply more than an nmtate interface.

I disagree, it was a documented behavior. As was the fact that all `*.yml` files in the directory would be applied. With the new change the very basis of the utility of the tool has changed. Let me do a better job of enumarating the problems:

1. one can no longer be assured that dropping a `.yml` in the directory and re-running `nmstatectl service` will apply the file
2. the contents of the `.applied` file can no longer be examined to see what state was being declared at the time of the last run
3.  Since all the `*.yml` files in the directory will be applied except for those with a `.applied` file that have a matching hash, there is no longer any guarantee that on a given run the files will be additively applied to arrive at a final declared state. e.g. if I have `01-rename-interfaces.yml`, `02-assemble-mlag-bonds.yml`, `03-apply-policy-based-routing.yml` a change made to any of the config files will no longer have been guaranteed to have been applied before subsequent configs are re-applied.

> There are other mechanism to check if the state was applied, like checking the `nmstatectl show` or if the nmstate.service unit was successful, but we are ok going back to old behaviour controller by a flag.

Yes, there are many ways to achieve various outcomes, but.... the existing documented behavior is depended upon for integration by myself and presumably others. Changing it in the manner it was changed is a breaking change to the documented and expected behavior.

I'm not saying that the behavior should never change or that the needs and intent of the maintainers should not drive development, but what I am saying is that version numbers should be meaningful and that existing documented behaviors/interfaces should not have changes of this nature made without relaying the change to the end users using the tools of documentation, version numbering, etc.

**Comment by qinqon** [comment]
> > @drawks we missed to document the change, although the .applied files were an implementation detail to prevent re-apply more than an nmtate interface.
> 
> I disagree, it was a documented behavior. As was the fact that all `*.yml` files in the directory would be applied. With the new change the very basis of the utility of the tool has changed. Let me do a better job of enumarating the problems:
> 
> 1. one can no longer be assured that dropping a `.yml` in the directory and re-running `nmstatectl service` will apply the file
> 2. the contents of the `.applied` file can no longer be examined to see what state was being declared at the time of the last run
> 3. Since all the `*.yml' files in the directory will be applied except for those with a `.applied`file that have a matching hash, there is no longer any guarantee that on a given run the files will be additively applied to arrive at a final declared state. e.g. if I have`01-rename-interfaces.yml`, `02-assemble-mlag-bonds.yml`, `03-apply-policy-based-routing.yml` a change made to any of the config files will no longer have been guaranteed to have been applied before subsequent configs are re-applied.
> 
> > There are other mechanism to check if the state was applied, like checking the `nmstatectl show` or if the nmstate.service unit was successful, but we are ok going back to old behaviour controller by a flag.
> 
> Yes, there are many ways to achieve various outcomes, but.... the existing documented behavior is depended upon for integration by myself and presumably others. Changing it in the manner it was changed is a breaking change to the documented and expected behavior.
> 
> I'm not saying that the behavior should never change or that the needs and intent of the maintainers should not drive development, but what I am saying is that version numbers should be meaningful and that existing documented behaviors/interfaces should not have changes of this nature made without relaying the change to the end users using the tools of documentation, version numbering, etc.

Right, we will be more careful next time, thanks for reporting.

**Comment by drawks** [comment]
> @drawks Are you seeking nmstate next release to change back to old behavior or expecting us to change the `CHANGLOG.md` and manpage?

Ideally the new behavior would be something that the user could opt into, the documentation would accurately describe both behaviors and if it is intended that the new behavior is going to be the default behavior at some future release there should be some form of deprecation warning in the docs and/or at runtime.

**Comment by tyll** [comment]
What's the reason to use a checksum in the first place? To me it seems that copying the file to `$file.applied` suffices, since then the original and the applied file can just be compared. This seems to reduce the dependencies, amount of code and also retains the ability to inspect what was applied. It will also be more matching a `remove_applied_state` configuration option because the change introduces two changes while the configuration option only mentions one change.

**Comment by tyll** [comment]
Also, the new dependency is a crypto library, so it is more likely to be subject to CVEs, therefore avoiding it has also most likely stronger benefits of not having to check for CVE issues because of it in the future.

**Comment by cathay4t** [comment]
> > @drawks Are you seeking nmstate next release to change back to old behavior or expecting us to change the `CHANGLOG.md` and manpage?
> 
> Ideally the new behavior would be something that the user could opt into, the documentation would accurately describe both behaviors and if it is intended that the new behavior is going to be the default behavior at some future release there should be some form of deprecation warning in the docs and/or at runtime.

Sure. I will work on PR for this opt-in.

**Comment by qinqon** [comment]
> Also, the new dependency is a crypto library, so it is more likely to be subject to CVEs, therefore avoiding it has also most likely stronger benefits of not having to check for CVE issues because of it in the future.

**Comment by qinqon** [comment]
> What's the reason to use a checksum in the first place? To me it seems that copying the file to `$file.applied` suffices, since then the original and the applied file can just be compared. This seems to reduce the dependencies, amount of code and also retains the ability to inspect what was applied. It will also be more matching a `remove_applied_state` configuration option because the change introduces two changes while the configuration option only mentions one change.

I think we are ok if we copy the file, the extra storage is not a big deal, also it means we are kind of retrocopatible with previous behaviour. 

@drawks are you ok if just copy verbatim the .yml to .applied ? this way you can keep checking it as before, and also the original file is not changed.

**Comment by cathay4t** [comment]
I have also create PR https://github.com/nmstate/nmstate/pull/2546 which allow user to opt-in the new behavior, but still by default delete/rename applied state files.

**Comment by drawks** [comment]
> > What's the reason to use a checksum in the first place? To me it seems that copying the file to `$file.applied` suffices, since then the original and the applied file can just be compared. This seems to reduce the dependencies, amount of code and also retains the ability to inspect what was applied. It will also be more matching a `remove_applied_state` configuration option because the change introduces two changes while the configuration option only mentions one change.
> 
> I think we are ok if we copy the file, the extra storage is not a big deal, also it means we are kind of retrocopatible with previous behaviour.
> 
> @drawks are you ok if just copy verbatim the .yml to .applied ? this way you can keep checking it as before, and also the original file is not changed.

I think maybe we are talking past one another a bit, but to be clear. I am not super in love with the 2.2.21 behavior, but it is the behavior that should be maintained as the default behavior. Which is to say the entire behavior as documented in the `nmstatect` manpage

```
       service
              Apply all network state files ending with .yml in specified( de‐
              fault:  /etc/nmstate)  folder.   The  applied network state file
              will be renamed with postfix .applied to  prevent  repeated  ap‐
              plied on next run.
```

and the `nmstate.service` manpage:

```
       nmstate.service invokes nmstatectl service command which apply all net‐
       work state files ending with .yml in /etc/nmstate folder.  The  applied
       network state file will be renamed with postfix .applied to prevent re‐
       peated applied on next service start.
```

As long as that is the default behavior I have no particular concern about the design of the new optional behavior. I do think that @tyll 's observation that copying the file instead of storing a hash is preferable for the reasons they have outlined.

Above all, I would like for a prolonged discussion about the design of this new optional behavior to not delay a release which reverts the change to the default behavior in its entirety, since it is currently shipping in the update repositories for the current stable 9.3 release of RHEL and RHEL clones/derivatives.


TLDR; please revert https://github.com/nmstate/nmstate/commit/dbb98e1d30c727b5e47ea7daacc8f2936ab5a0e0 and cut a new release, the changed behavior should never have been the default and a new behavior being introduced as optional is an entirely unrelated concern to the need to fix the currently broken release.

**Comment by drawks** [comment]
> I have also create PR #2546 which allow user to opt-in the new behavior, but still by default delete/rename applied state files.

I have added a comment to the discussion of your MR along with a specific example of a use case where the 2.2.21 behavior could be used in accordance with the documented behavior and a changed behavior would lead to negative outcomes.

https://github.com/nmstate/nmstate/pull/2546#issuecomment-1932515525

I very much appreciate the attention that is being given to this issue, but my primary concern is to not introduce ANY change that breaks the interface of the 2.2.21 release.

This software does have a 3 part version number which generally implies some amount of semantic versioning:

1. MAJOR version when you make incompatible API changes
2. MINOR version when you add functionality in a backward compatible manner
3. PATCH version when you make backward compatible bug fixes

introducing new non-breaking changes should happen in minor releases and breaking changes should happen in major releases

**Comment by sq-gh** [comment]
> > I have also create PR #2546 which allow user to opt-in the new behavior, but still by default delete/rename applied state files.
> 
> I have added a comment to the discussion of your MR along with a specific example of a use case where the 2.2.21 behavior could be used in accordance with the documented behavior and a changed behavior would lead to negative outcomes.
> 
> [#2546 (comment)](https://github.com/nmstate/nmstate/pull/2546#issuecomment-1932515525)
> 
> I very much appreciate the attention that is being given to this issue, but my primary concern is to not introduce ANY change that breaks the interface of the 2.2.21 release.
> 
> This software does have a 3 part version number which generally implies some amount of semantic versioning:
> 
>     1. MAJOR version when you make incompatible API changes
> 
>     2. MINOR version when you add functionality in a backward compatible manner
> 
>     3. PATCH version when you make backward compatible bug fixes
> 
> 
> introducing new non-breaking changes should happen in minor releases and breaking changes should happen in major releases

With nmstate aimed to satisfy enterprise needs, following semantic versioning spec is essential. 

From semver:
> Bug fixes not affecting the API increment the patch version, backward compatible API additions/changes increment the minor version, and backward incompatible API changes increment the major version.

The next point release should revert this breaking change.

**Comment by cathay4t** [comment]
I think the point is not about whether @drawks is OK with #2545 , but maintain API stable.

I will rebase my PR #2546 over #2545.

**Comment by cathay4t** [comment]
The 2.2.24 release should fixed your problem. Please check again.

**Comment by drawks** [comment]
> The 2.2.24 release should fixed your problem. Please check again.

It doesn't seem to have been released into the rhel 9.3 appstream repo yet for me to redo integration testing

**Comment by cathay4t** [comment]
It will take some weeks for 9.3 release.

**Comment by cathay4t** [comment]
9.3 got the update yesterday(2024 March 03), please check again.

**Comment by drawks** [comment]
I've tested the most recent RHEL packaged release and agree that this issue seems resolved. Thanks everyone!
